### ISSUE TITLE
EMEA has 0/NaN value checkpoint not being saved

### ISSUE THREAD

**Issue by rajitjijo** [ultralytics/ultralytics#24615]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I would have attached a screenshot but it keeps on saying failed to upload, but regardless, i have been fine tuning the yolo11 model and ive been trying to use the AdamW as my optimizer but i keep getting the warning "Skipping Checkpoint save at epoch since EMA contains NaN/inf" I was following the threads and i saw the AdamW was fixed, so i tried to use it.

I do get numerical values for losses and accuracies, but no weights are being saved because of that warning.

Here are my hyperparameters for training:
```
lr0: 0.0005
lrf: 0.01
momentum: 0.843
weight_decay: 0.00036
warmup_epochs: 1.0
warmup_momentum: 0.5
warmup_bias_lr: 0.05
box: 0.0296
cls: 0.243
hsv_h: 0.0138
hsv_s: 0.664
hsv_v: 0.464
degrees: 0.373
flipud: 0.00856
fliplr: 0.5
mosaic: 0.8
mixup: 0.05
copy_paste: 0.0
scale: 0.5
shear: 0.1
```

### Environment

Ultralytics 8.4.51 🚀 Python-3.14.4 torch-2.12.0+cu130 CUDA:0 (NVIDIA RTX PRO 6000 Blackwell Server Edition, 97252MiB)
Setup complete ✅ (8 CPUs, 62.3 GB RAM, 563.3/1938.9 GB disk)

OS                     Linux-7.0.0-1004-aws-x86_64-with-glibc2.43
Environment            Linux
Python                 3.14.4
Install                pip
Path                   /home/ubuntu/yolov26/yolo_env/lib/python3.14/site-packages/ultralytics
RAM                    62.27 GB
Disk                   563.3/1938.9 GB
CPU                    Intel Xeon Platinum 8559C
CPU count              8
GPU                    NVIDIA RTX PRO 6000 Blackwell Server Edition, 97252MiB
GPU count              1
CUDA                   13.0

numpy                  ✅ 2.4.5>=1.23.0
matplotlib             ✅ 3.10.9>=3.3.0
opencv-python          ✅ 4.13.0.92>=4.6.0
pillow                 ✅ 12.2.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.34.2>=2.23.0
scipy                  ✅ 1.17.1>=1.4.1
torch                  ✅ 2.12.0>=1.8.0
torch                  ✅ 2.12.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.27.0>=0.9.0
psutil                 ✅ 7.2.2>=5.8.0
polars                 ✅ 1.40.1>=0.20.0
ultralytics-thop       ✅ 2.0.19>=2.0.18

### Minimal Reproducible Example

```
from ultralytics import YOLO
import torch
import yaml
from datetime import datetime

model = YOLO("runs/detect/Training_Run_YOLO11_05-18-2026-22-30/train/weights/best.pt")
device = "cuda"
model.to(device)


if __name__ == "__main__":

    project_name = f'Training_Run_YOLO11_{datetime.now().strftime("%m-%d-%Y-%H-%M")}'

    model.train(data="pronto_data.yaml", imgsz=640, batch=96, 
                epochs=50, patience=15,
                cfg="hyperparameters.yaml",
                project=project_name,
                optimizer="AdamW",
                device=0,
                workers=2,
                )
```

the weights are yolo11.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rajitjijo, thank you for reporting this and for providing a detailed environment summary plus a reproducible training snippet 🚀 This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, thank you for already including a reproducible example. To help with debugging, please also share any full training log output around the `Skipping Checkpoint save at epoch since EMA contains NaN/inf` warning, and confirm whether this reproduces after upgrading to the latest package and dependencies.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by lmycross** [comment]
Can you try disabling AMP by setting `amp=False`, or use `optimizer="SGD"` or `optimizer="MuSGD"` instead?

**Comment by kim-evan-kim-GitLB** [comment]
I am also currently using a Pro 6000, and I am experiencing the same issue where the EMA is corrupted and the data points are not being saved.

This problem has been occurring since updating to version 8.4.48.

I am currently attempting to train the model by downgrading to version 8.4.37. If the issue persists with the same parameters in that version, I will leave an additional comment.

I plan to conduct experiments using the same parameters and dataset as in 8.4.48.

Is this a problem with the training parameters? Why is this happening?

**Comment by glenn-jocher** [comment]
This does not look like a normal hyperparameter issue; the save is being skipped only when the EMA weights contain non-finite values, so this is more consistent with a numerical stability or environment regression than with your dataset settings alone. Please post a text MRE using `yolo26n.pt` + `coco8.yaml`, the full first warning/log output, and the result of `from ultralytics.utils.checks import collect_system_info; collect_system_info()` so we can verify the exact loaded package path and stack, and please compare against the [AMP check behavior](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp) and our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum-reproducible-example/).

**Comment by Chung-Lucy** [comment]
I encountered the same problem with Ultralytics versions 8.4.39 and 8.4.59 when training on my dataset.

model.train(
    data="animal.yaml",
    imgsz=640,
    batch=32,
    epochs=120,
    patience=15,
    project=project_name,
    optimizer="AdamW",
    device=0,
    workers=12,
    lr0=0.01,
    warmup_epochs=5,
    perspective=0.001,
    mosaic=0.6,
    ...  # other settings are the same as default
)

The issue can be resolved by setting "--no-amp", but the mAP drops by 1%, and training stops early at epoch 17.

However, after downgrading to Ultralytics versions 8.3.105 and 8.3.202, training runs normally now.

**Comment by glenn-jocher** [comment]
Thanks, that helps narrow this down to an AMP-related regression in the newer stack rather than a dataset tuning issue; please share a text repro with `yolo26n.pt` + `coco8.yaml`, the first warning/complete traceback, and `from ultralytics.utils.checks import collect_system_info; collect_system_info()` from the exact loaded install so we can verify whether [AMP checks](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp) should be disabling AMP earlier or whether non-finite values are first appearing in the [checkpoint save path](https://docs.ultralytics.com/reference/engine/trainer#ultralytics.engine.trainer.BaseTrainer.save_model).

**Comment by Chung-Lucy** [comment]
[1] yolo26n.pt + coco8.yaml
```
from ultralytics import YOLO

model = YOLO("yolo26n.pt")

model.train(
    data="coco8.yaml",
    epochs=120,
    imgsz=640,
    batch=32,
    device=[0],
    workers=12,
    optimizer="AdamW",
    lr0=0.001,
    lrf=0.05,
    warmup_epochs=5,
    box=0.05,
    degrees=10,
    perspective=0.001,
    fliplr=0,
    mosaic=0.6,
    mixup=0.2,
    auto_augment=None,
    freeze=0,  
    patience=15,
    cache="disk"
)
```
[2] "Generally, the warning started occurring at epoch 17 and persisted until the training finished:
WARNING ⚠️ Skipping checkpoint save at epoch 17: EMA contains NaN/Inf"

[3] System Info
Ultralytics 8.4.39 🚀 Python-3.10.20 torch-2.4.0 CUDA:0 (NVIDIA GeForce RTX 4090, 24210MiB)
Setup complete ✅ (32 CPUs, 125.5 GB RAM, 693.4/915.3 GB disk)

OS                     Linux-6.8.0-64-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.10.20
Install                pip
Path                   /home2/user/lucychung/miniforge3/envs/Pet-Cls/lib/python3.10/site-packages/ultralytics
RAM                    125.51 GB
Disk                   693.4/915.3 GB
CPU                    Intel Core i9-14900
CPU count              32
GPU                    NVIDIA GeForce RTX 4090, 24210MiB
GPU count              2
CUDA                   12.1

numpy                  ✅ 2.0.2>=1.23.0
matplotlib             ✅ 3.10.8>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 12.2.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.34.2>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.4.0>=1.8.0
torch                  ✅ 2.4.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.19.0>=0.9.0
psutil                 ✅ 7.2.2>=5.8.0
polars                 ✅ 1.41.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

[4] check amp
```
AMP: running Automatic Mixed Precision (AMP) checks...
AMP: checks passed  ✅ 
```

**Comment by glenn-jocher** [comment]
Thanks, this is helpful — a passing [AMP check](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp) only verifies short FP32 vs AMP inference, while your warning is coming later from the [checkpoint save path](https://docs.ultralytics.com/reference/engine/trainer#ultralytics.engine.trainer.BaseTrainer.save_model) when EMA already contains non-finite tensors, so this still looks like training-time AMP instability. Since the repro is on `ultralytics==8.4.39` with `torch==2.4.0`, please retest on the current `ultralytics` release with a newer PyTorch build and, if it still reproduces, attach the `results.csv` rows around epoch 17 so we can see which metric goes non-finite first.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24731: Prevent NaN/Inf EMA from discarding training checkpoints](https://github.com/ultralytics/ultralytics/pull/24731) by @glenn-jocher.

This change may resolve the issue by improving how training checkpoints and validation handle FP16 overflow edge cases:

- Checkpoint saving in `trainer.py` was updated so checkpoints are no longer skipped when EMA weights are still valid in FP32 but overflow only during FP16 serialization.
- A safeguard now clamps non-finite FP16 values on the serialized checkpoint copy using `torch.nan_to_num_()`, without modifying the live EMA model in memory.
- The checkpoint validity check now inspects the original FP32 EMA weights first, instead of only checking after FP16 conversion.
- Validation in `validator.py` now keeps weights in FP32 during training-time validation, while still using autocast where appropriate for inference and loss.
- A regression test was added to confirm checkpoints are still saved in this scenario, remain finite on disk, and do not corrupt the live EMA model.

The PR also bumps the package version to `8.4.62`.

Credit to @glenn-jocher for the fix and to all contributors who helped investigate and improve this behavior.

If you'd like to test the changes, you can try either:

- `pip install -U ultralytics>=8.4.62`
- `pip install git+https://github.com/ultralytics/ultralytics.git@main`

If `8.4.62` is not yet available on PyPI in your environment, testing from `main` is the best option, or you can wait for the next official release.

Please let us know whether this resolves the issue on your side. Thanks 🙏 for reporting this, and any additional feedback is welcome if the issue still persists.
